src/workers/block_listenser.rs

```diff
@@ -296,94 +294,60 @@ impl BlockListener {
-            });
-        }
+            let mut slot_que = Vec::new();
 
-        let rpc_client = self.rpc_client.clone();
-        tokio::spawn(async move {
-            let slots_task_queue = slots_task_queue.clone();
-            let last_latest_slot = self
-                .block_store
-                .get_latest_block_info(commitment_config)
-                .await
-                .slot;
-            // -5 for warmup
-            let mut last_latest_slot = last_latest_slot - 5;
-
-            // storage for recent slots processed
-            let rpc_client = rpc_client.clone();
             loop {
-                let new_slot = rpc_client
-                    .get_slot_with_commitment(commitment_config)
+                let mut new_block_slots = self
+                    .rpc_client
+                    .get_blocks_with_commitment(latest_slot, None, commitment_config)
                     .await?;
 
-                if last_latest_slot == new_slot {
-                    tokio::time::sleep(tokio::time::Duration::from_millis(200)).await;
-                    println!("no slots");
+                if new_block_slots.is_empty() {
+                    warn!("{latest_slot} No slots");
+                    tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
                     continue;
                 }
 
-                // filter already processed slots
-                let new_block_slots: Vec<u64> = (last_latest_slot..new_slot).collect();
-                // context for lock
-                {
-                    let mut lock = slots_task_queue.lock().await;
-                    for slot in new_block_slots {
-                        lock.push_back((slot, 0));
-                    }
+                let new_latest_slot = *new_block_slots.last().unwrap();
+
+                if latest_slot == new_latest_slot {
+                    warn!("No new slots for {latest_slot}");
+                    tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
+                    continue;
                 }
 
-                last_latest_slot = new_slot;
-                tokio::time::sleep(tokio::time::Duration::from_millis(200)).await;
+                latest_slot = new_latest_slot;
+
+                // reverse to put latest_slot first
+                new_block_slots.reverse();
+
+                // need not queue up new slots if they are less than 16 and there the queue is empty
+                let slots_to_get_blocks = if slot_que.is_empty() && new_block_slots.len() <= 16 {
+                    new_block_slots
+                } else {
+                    slot_que.append(&mut new_block_slots);
+                    slot_que.split_off(slot_que.len().min(16))
+                };
+
+                let index_futs = slots_to_get_blocks
+                    .into_iter()
+                    .map(|slot| self.index_slot(slot, commitment_config, postgres.clone()));
+
+                join_all(index_futs).await;
```

Comment:
Before we retried if the fetch of block failed for some reason. You should implement a retry logic.

Reply:
Fixed in 94ad00d805b15b97950d227d35b2e26536b1da5f